text/0069-api-v2.md

```diff
@@ -0,0 +1,225 @@
+
+#### RawKV Value Encode
+
+If the key has RawKV prefix, which are `k{keyspace id prefix}r` or `k{keyspace id prefix}a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+#### TTL in Titan
+
+Titan is a storage engine plugin for RocksDB that store the value of a kvpair in a separated blob file if the value is large enough. This brings a new challenge for RawKV TTL compaction filter where the compaction filter can not know the ttl without reading the blob file. So with Titan, the RawKV value encode is defined to be:
+
+##### Value
+
+1. `{0x0}`
+2. `{0x1}{TTL expire timestamp}`
+
+##### Blob Value
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### Keyspace Management
+
+Add a new http interface to PD for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    id: "0"
+  },
+  {
+    name: "redis",
+    id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "foo",
+}
+
+// delete a keyspace
+DELETE /keyspaces/{keyspace_name}
+{}
+
+// recover the latest deleted keyspace
+POST /keyspaces?action=flashback
+{
+  new_name: "bar",
+}
+```
+
+The keyspaces are stored in etcd and has the no limitation on the id number.
+
+1. Adding keyspace: newly added keyspaces can only be viable to new clients.
+2. Deleting keyspace:
+  2.1. Deleting keyspace only marks the metadata to inviable in PD, the data in the keyspace and metadata in PD will not be deleted automatically. Garbage collecting the data in deleted keyspaces may be introduced in the future since it increases the complexity of this RFC. To ensure no data is left, the user should clean up the keyspace before deleting the keyspace at present.
+  2.2. Every client syncs the keyspace information with the PD leader every 5 minutes. When a keyspace is deleted, the client should be aware of the deletion in 5 minutes. The keyspace can not be accessed by any living client after 5 minutes.
+3. Flashbacking keyspace:
+  3.1. Flashbacking keyspace only affects the metadata, turning the keyspace name and id mapping to viable by clients.
+  3.2. If there are multiple deleted keyspaces with the same name, only the last deleted keyspace is flashbacked. Users can flashback all these deleted keyspaces by calling the flashback API multiple times with different new keyspace names.
+
+#### pd-ctl
+
+To enable API V2, which also enables the keyspace API:
+
+```bash
+>> config set api-version 2
+```
+
+To manage keyspace:
+
+```bash
+>> config keyspaces show
+>> config keyspaces create <keyspace name>
+>> config keyspaces delete <keyspace name>
+>> config keyspaces flashback <new keyspace name>
+```
+
+#### Keyspace metadata
+
+```json
+{
+  name: string,
+  id: int64,
+  created_at: timestamp,
+  deleted_at: timestamp, // if set, the keyspace is not visiable to users
+  flashbacked_at: timstamp
+}
```

Comment:
Can we extend the metqdata？ e.g. add a new filed `default_TTL`, which means if users put a kv without ttl, the `defualt_TTL` will be used.

Reply:
LGTM